apps/hash-ai-worker-ts/src/activities/shared/get-llm-response.ts

```diff
@@ -666,15 +689,170 @@ const getOpenAiResponse = async (
  * `model` provided in the parameters.
  */
 export const getLlmResponse = async <T extends LlmParams>(
-  params: T,
+  params: T & {
+    /**
+     * Required for tracking usage on a per-user basis.
+     *
+     * @todo: consider abstracting this in a wrapper method, or via
+     * generic params (via a `logUsage` method).
+     */
+    userAccountId: AccountId;
+    graphApiClient: GraphApi;
+    linkUsageRecordToEntities: {
+      linkEntityTypeId: [
+        typeof systemLinkEntityTypes.incurredIn.linkEntityTypeId,
+        typeof systemLinkEntityTypes.created.linkEntityTypeId,
+        typeof systemLinkEntityTypes.updated.linkEntityTypeId,
+      ][number];
+      entityId: EntityId;
+    }[];
+  },
 ): Promise<LlmResponse<T>> => {
-  if (isLlmParamsAnthropicLlmParams(params)) {
-    const response = await getAnthropicResponse(params);
+  const {
+    userAccountId,
+    graphApiClient,
+    linkUsageRecordToEntities,
+    ...remainingLlmParams
+  } = params;
 
-    return response as unknown as LlmResponse<T>;
-  } else {
-    const response = await getOpenAiResponse(params);
+  /**
+   * Check whether the user has exceeded their usage limit, before
+   * proceeding with the LLM request.
+   */
+  const userHasExceededUsageStatus =
+    await userExceededServiceUsageLimitActivity({
+      graphApiClient,
+      userAccountId,
+    });
+
+  if (userHasExceededUsageStatus.code !== StatusCode.Ok) {
+    return {
+      status: "exceeded-usage-limit",
+      message:
+        userHasExceededUsageStatus.message ??
+        "You have exceeded your usage limit.",
+    };
+  }
 
-    return response as unknown as LlmResponse<T>;
+  let aiAssistantAccountId: AccountId;
+  try {
+    aiAssistantAccountId = await getMachineActorId(
+      { graphApi: graphApiClient },
+      { actorId: userAccountId },
+      { identifier: "hash-ai" },
+    );
+  } catch {
+    return {
+      status: "internal-error",
+      message: `Failed to retrieve AI assistant account ID ${userAccountId}`,
+    };
   }
+
+  /**
+   * @todo: figure out why TS is not inferring the type of `remainingLlmParams`
+   * correctly (may not be relevant once usage tracking related params are removed)
+   */
+  const llmParams = remainingLlmParams as unknown as T;
```

Comment:
I would split them out so you don't have to rely on a `...rest`, e.g.
```ts
export const getLlmResponse = async <T extends LlmParams>(
  llmParams: T,
  usageTracking: {
    //etc
  }
)
```
This solves the type problem and maintains a clearer separation between 'stuff for the LLM call' and 'other internal stuff'

Reply:
Good idea, much prefer this - have done so in https://github.com/hashintel/hash/pull/4379/commits/9fd6cc2de1f0da9bee00171b6f06cade48ca0075